artemis/reporting/utils.py

```diff
@@ -0,0 +1,122 @@
+import functools
+import urllib.parse
+from socket import gethostbyname, getservbyport
+from typing import Any, Dict, Optional
+
+from artemis import http_requests
+
+
+def get_target(task_result: Dict[str, Any]) -> str:
+    """Returns a string representation of the target that has been scanned."""
+    headers = task_result["headers"]
+    payload = task_result["payload"]
+    if "url" in payload:
+        assert isinstance(payload["url"], str)
+        return payload["url"]
+    if "host" in payload and headers["service"] == "http":
+        assert isinstance(payload["host"], str)
+        return ("https" if payload["ssl"] else "http") + "://" + payload["host"] + ":" + str(payload["port"]) + "/"
+    raise NotImplementedError(f"Unable to obtain viable target in {payload}")
+
+
+def get_top_level_target_if_present(task_result: Dict[str, Any]) -> Optional[str]:
+    """Returns the top level target - i.e. what was initially provided by the user as the target to scan.
+
+    For example, the top level target may be example.com, but the actual vulnerability may be
+    found on https://subdomain.example.com:443/.
+    """
+    payload_persistent = task_result["payload_persistent"]
+
+    if "original_domain" in payload_persistent:
+        assert isinstance(payload_persistent["original_domain"], str)
+        return payload_persistent["original_domain"]
+    elif "original_ip" in payload_persistent:
+        assert isinstance(payload_persistent["original_ip"], str)
+        return payload_persistent["original_ip"]
+    else:
+        return None
+
+
+def get_top_level_target(task_result: Dict[str, Any]) -> str:
+    target = get_top_level_target_if_present(task_result)
+    if target:
+        return target
+    else:
+        raise ValueError(f"No top level target found in {task_result}")
+
+
+def get_scheme_from_url(url: str) -> str:
+    return urllib.parse.urlparse(url).scheme
+
+
+def get_host_from_url(url: str) -> str:
+    host = urllib.parse.urlparse(url).hostname
+    assert host is not None
+    return host
+
+
+def get_port_from_url(url: str) -> int:
+    url_parsed = urllib.parse.urlparse(url)
+    port = url_parsed.port
+
+    if not port:
+        if url_parsed.scheme == "http":
+            return 80
+        elif url_parsed.scheme == "https":
+            return 443
+        else:
+            raise NotImplementedError()
+    return port
+
+
+@functools.lru_cache(maxsize=8192)
+def cached_gethostbyname(host: str) -> str:
+    """Please keep in mind that the requests will not be proxied
+    and will leak information about what you are testing."""
+    return gethostbyname(host)
```

Comment:
We have this in `artemis.utils`. Don't duplicate code. If `lru_cached` version is required, feel free to add it there.

Don't use `gethostbyname`.

Reply:
This is cached version of gethostbyname. I thought you are ok with http or dns requests from here if we write that in docs?

Reply:
If you want to refactor autoreporter to make all requests proxable, feel free to add a task to do that, but in a separate PR.

Reply:
No. There are ready utilities. In place. To be used. This method gives absolutely **no** advantages over existing implementation, but duplicates code. If you want a cached version, implement it in artemis.utils modules. Perhaps even the global version could use lru_cache.